Add draft for EOF Functions (#4750)
ethereum/EIPs: EIPS/eip-4750.md

status: Draft
type: Standards Track
category: Core
created: 2022-01-10
requires: 3540, 3670
---

## Abstract

Introduce the ability to have several code sections in EOF-formatted ([EIP-3540](./eip-3540.md)) bytecode, each one representing a separate subroutine/function. Two new opcodes,`CALLF` and `RETF`, are introduced to call and return from such a function.

## Motivation

Currently in the EVM everything is a dynamic jump. Languages like Solidity generate most jumps in a static manner (i.e. the destination is pushed to the stack right before, `PUSHn .. JUMP`). Unfortunately however this cannot be used by most EVM interpreters, because of added requirement of validation/analysis. This also restricts them from making optimisations and potentially reducing the cost of jumps.

[EIP-4200](./eip-4200.md) introduces static jump instructions, which remove the need for *most* dynamic jump use cases, but not everything can be solved with them.

This EIP aims to remove the need for dynamic jumps as it offers the most important feature those are used for: calling into and returning from functions. While it removes the need, it does not disallow those instructions.

> Furthermore it aims to improve analysis opportunities by encoding the number of intputs and outputs for each given function, and isolating the stack of each function (i.e. a function cannot read the stack of the caller/callee).

Reviewer: ```suggestion
Furthermore it aims to improve analysis opportunities by encoding the number of inputs and outputs for each given function, and isolating the stack of each function (i.e. a function cannot read the stack of the caller/callee).
```

1. Execution starts at the first byte of the first code section, and PC is set to 0.
2. Return stack is initialized to contain one item: `(code_section_index = 0, offset = 0, stack_height = 0)`
3. Destinations of jumps are allowed only to be inside current code section. `JUMP` and `JUMPI` result in exceptional halt when destination is outside of current section bounds.
4. If any instruction would access a data stack item below `caller_stack_height`, execution results in exceptional halt. This rule replaces the old stack underflow check.

#### Implications on the JUMPDEST analysis

- Analysis is done separately for each section, i.e. output of entire analysis is `number_of_code_sections` lists of possible jump destinations.
- Analysis is extended to consider 2 bytes directly following `CALLF` to be invalid jump destination

## Rationale

### `RETF` in the top frame ends execution vs exceptionally halts

Alternative logic for executing `RETF` in the top frame could be to exceptionally halt execution, because there is arguably no caller for the starting function. This would mean that return stack is initialized as empty, and `RETF` exceptionally aborts when return stack is empty.

> We have decided in favor of always having at least one item in the return stack, becasue it allows to avoid having a special case for empty stack in the interprer loop stack underflow check. We keep the stack underflow rule general by having `caller_stack_height = 0` in the top frame.

Reviewer: ```suggestion
We have decided in favor of always having at least one item in the return stack, because it allows to avoid having a special case for empty stack in the interpreter loop stack underflow check. We keep the stack underflow rule general by having `caller_stack_height = 0` in the top frame.
```

Author: Oh come on, codespell should have caught this :)

2. EIP-3670 requires that immediate arguments of instructions (`PUSHn`, `RJUMP`, `RJUMPI`) are inside code section bounds. This rule is extended to include `CALLF` into the list of such instructions. That is, 2 bytes directly following `CALLF` must be inside code section.

> TODO In case EIP-3670 includes terminator instruction requirement:
> 3. List of allowed terminating instructions is extended to include `RETF`.
> (rule 2 above is not really needed then, is is implied from this)
3. Code section is invalid in case an immediate argument of any `CALLF` is greater than or equal to the total number of code sections.
4. `RJUMP` and `RJUMPI` immediate argument value (jump destination relative offset) validation:
    4.1. Code section is invalid in case offset points to a position outside of section bounds.
    4.2. Code section is invalid in case offset points to one of two bytes directly following `CALLF` instruction.

### Execution

1. Execution starts at the first byte of the first code section, and PC is set to 0.
2. Return stack is initialized to contain one item: `(code_section_index = 0, offset = 0, stack_height = 0)`

> 3. Jumps destinations are allowed only to be inside current code section. `JUMP` and `JUMPI` result in exceptional halt when destination is outside of current section bounds.

Reviewer 2: Maybe better: destinations of jump opcodes (to distinguish this from the jumpdest opcode)

Author: Changed.

> 4. If any instruction would change the stack to height below `caller_stack_height`, execution results in exceptional halt. This rule replaces the old stack underflow check.

Reviewer 2: ```suggestion
4. If any instruction would change the stack to height below `caller_stack_height` or accesses a data stack item below that height, execution results in exceptional halt. This rule replaces the old stack underflow check.
```

Reviewer 2: Explanation: It can be argued if `swap` instructions actually change the stack height.

Author: Good point!

Author: Changed.
